spec/terrafying/components/vpc_spec.rb

```diff
@@ -0,0 +1,207 @@
+
+      original_route_count = vpc_a.output_with_children["resource"]["aws_route"].count
+
+      vpc_a.peer_with(vpc_b)
+
+      num_new_routes = vpc_a.output_with_children["resource"]["aws_route"].count - original_route_count
+
+      expect(num_new_routes).to eq(vpc_a.subnets.count * vpc_a.azs.count + vpc_b.subnets.count * vpc_b.azs.count)
+    end
+
+    it "should allow users to limit the subnets that are peered" do
+      vpc_a = Terrafying::Components::VPC.create("a", "10.0.0.0/16")
+      vpc_b = Terrafying::Components::VPC.create("b", "10.1.0.0/16")
+
+      original_route_count = vpc_a.output_with_children["resource"]["aws_route"].count
+
+      vpc_a.peer_with(vpc_b, { our_subnets: [vpc_a.subnets[:public]],
+                               their_subnets: [vpc_b.subnets[:public]] })
+
+      num_new_routes = vpc_a.output_with_children["resource"]["aws_route"].count - original_route_count
+
+      expect(num_new_routes).to eq(1 * vpc_a.azs.count + 1 * vpc_b.azs.count)
+    end
+
+  end
+
+  context "extract_subnet!" do
+
+    it "should limit the size of a subnet to a /28" do
+      vpc = Terrafying::Components::VPC.create("foo", "10.0.0.0/16")
+      cidr = vpc.extract_subnet!(30)
+
+      expect(cidr).to match(/[\.0-9]+\/28/)
+    end
+
+    it "should raise when there are no subnets left" do
+      vpc = Terrafying::Components::VPC.create("foo", "10.0.0.0/16")
+
+      249.times {
+        vpc.extract_subnet!(24)
+      }
+
+      expect {
+        vpc.extract_subnet!(24)
+      }.to raise_error(RuntimeError)
+    end
+
+  end
+
+  context "allocate_subnet!" do
+
+    it "should create subnets for each availability zone" do
+      vpc = Terrafying::Components::VPC.create("foo", "10.0.0.0/16")
+      subnets = vpc.allocate_subnets!("asdf")
+
+      expect(subnets.count).to eq(vpc.azs.count)
+    end
+
+    it "should attach an internet gateway if subnet is public" do
+      vpc = Terrafying::Components::VPC.create("foo", "10.0.0.0/16")
+      subnets = vpc.allocate_subnets!("asdf", { public: true })
+
+      output = vpc.output_with_children
+
+      route = output["resource"]["aws_route"].values.select { |route|
+        route[:route_table_id] = subnets[0].route_table
+      }.first
+
+      expect(route.has_key?(:gateway_id)).to be true
+      expect(route.has_key?(:nat_gateway_id)).to be false
+    end
+
+    it "should attach a NAT gateway if it's connected to the internet but not public" do
+      vpc = Terrafying::Components::VPC.create("foo", "10.0.0.0/16")
+      subnets = vpc.allocate_subnets!("asdf", { public: false, internet: true })
+
+      output = vpc.output_with_children
+
+      route = output["resource"]["aws_route"].values.select { |route|
+        route[:route_table_id] == subnets[0].route_table
+      }.first
+
+      expect(route.has_key?(:gateway_id)).to be false
+      expect(route.has_key?(:nat_gateway_id)).to be true
+    end
+
+    it "should not have any routes if it isn't public and no internet" do
```

Comment:
Just commenting to make sure this test gets filled in later :) 👍

Reply:
Hahaha, you caught me

Reply:
:shipit:

Reply:
🥇